models/glif_psc.cpp

```diff
@@ -0,0 +1,578 @@
+  {
+    throw BadProperty( "Bad glif model type string." );
+  }
+  long model_type = nest::glif_psc::model_type_lu[ model_str ];
+
+  double v_old = S_.U_;
+  double spike_component = 0.0;
+  double voltage_component = 0.0;
+  double th_old = S_.threshold_;
+
+  for ( long lag = from; lag < to; ++lag )
+  {
+
+    // update threshold via exact solution of dynamics of spike component of
+    // threshold
+    // exact solution of dynamics of spike component of threshold for glif2/4/5
+    // models with "R"
+    if ( model_type == 2 || model_type == 4 || model_type == 5 )
+    {
+      spike_component = V_.last_spike_ * std::exp( -P_.b_spike_ * dt );
+    }
+    V_.last_spike_ = spike_component;
+
+    if ( V_.t_ref_remaining_ > 0.0 )
+    {
+      // While neuron is in refractory period count-down in time steps (since dt
+      // may change while in refractory) while holding the voltage at last peak.
+      V_.t_ref_remaining_ -= dt;
+      if ( V_.t_ref_remaining_ <= 0.0 )
+      {
+        // Neuron has left refractory period, reset voltage and after-spike
+        // current
+
+        // Reset ASC_currentsfor glif3/4/5 models with "ASC"
+        if ( model_type == 3 || model_type == 4 || model_type == 5 )
+        {
+          for ( std::size_t a = 0; a < S_.ASCurrents_.size(); ++a )
+          {
+            S_.ASCurrents_[ a ] =
+              P_.asc_amps_[ a ] + S_.ASCurrents_[ a ] * P_.r_[ a ] * std::exp( -P_.k_[ a ] * V_.t_ref_total_ );
+          }
+        }
+
+        // Reset voltage
+        if ( model_type == 1 || model_type == 3 )
+        {
+          // Reset voltage for glif1/3 models without "R"
+          S_.U_ = P_.V_reset_;
+        }
+        else
+        {
+          // Reset voltage for glif2/4/5 models with "R"
+          S_.U_ = P_.voltage_reset_a_ * S_.U_ + P_.voltage_reset_b_;
+
+          // reset spike component of threshold
+          V_.last_spike_ = V_.last_spike_ + P_.a_spike_;
+
+          // rest the global threshold (voltage component of threshold: stay the
+          // same)
+          S_.threshold_ = V_.last_spike_ + V_.last_voltage_ + P_.th_inf_;
+        }
+        // Check if bad reset
+        // TODO: Better way to handle?
+        if ( S_.U_ > S_.threshold_ )
+        {
+          printf(
+            "Simulation Terminated: Voltage (%f) reset above threshold "
+            "(%f)!!\n",
+            S_.U_,
+            S_.threshold_ );
+        }
+        assert( S_.U_ <= S_.threshold_ );
+      }
+      else
+      {
+        S_.U_ = v_old;
+      }
+    }
+    else
+    {
+      // Integrate voltage and currents
+
+      // Calculate new ASCurrents value using expoential methods
+      S_.ASCurrents_sum_ = 0.0;
+      // for glif3/4/5 models with "ASC"
+      if ( model_type == 3 || model_type == 4 || model_type == 5 )
+      {
+        for ( std::size_t a = 0; a < S_.ASCurrents_.size(); ++a )
+        {
+          S_.ASCurrents_sum_ += S_.ASCurrents_[ a ];
```

Comment:
Why do you sum before updating?

Reply:
The ASC sum is used to compute the Vm value of next step for the dynamics. The ASC updating is in the same manner to compute the ASC for the next step.

Reply:
Ok. You sum the currents at the beginning of the time step and take that value as constant throughout the time step; and then you update the current values to the value at the end of the time step, ready for the next time step. You may want to comment this. Strictly speaking, the ASC should be included in the exact integration scheme as well, but that would be challenging given that the number of currents can be changed by the user. Are the time constants reasonably long to limit integration error?